src/fastpath/fastpath.c

```diff
@@ -902,3 +902,252 @@ void NORETURN fastpath_vm_fault(vm_fault_type_t type)
+        mask_ack_bail(irq);
+        UNREACHABLE();
+    }
+
+    if (unlikely(intStateIRQTable[irq] != IRQSignal)) {
+        slowpath_irq();
+        UNREACHABLE();
+    }
+
+    cap_t ntfn_cap = intStateIRQNode[irq].cap;
+    if (unlikely(cap_get_capType(ntfn_cap) != cap_notification_cap ||
+                !cap_notification_cap_get_capNtfnCanSend(ntfn_cap))) {
+        if (config_set(CONFIG_IRQ_REPORTING)) {
+            printf("Undelivered irq: %d\n", (int) irq);
+        }
+        mask_ack_bail(irq);
+        UNREACHABLE();
+    }
+
+    notification_t *ntfn_ptr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(ntfn_cap));
+    notification_state_t ntfn_state = notification_ptr_get_state(ntfn_ptr);
+    word_t badge = cap_notification_cap_get_capNtfnBadge(ntfn_cap);
+
+    switch (ntfn_state) {
+    case NtfnState_Active:
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfn_ptr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfn_ptr));
+        mask_ack_bail(irq);
+        UNREACHABLE();
+    case NtfnState_Idle:
+        dest = (tcb_t *) notification_ptr_get_ntfnBoundTCB(ntfn_ptr);
+
+        if (!dest || thread_state_ptr_get_tsType(&dest->tcbState) != ThreadState_BlockedOnReceive) {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            ntfn_set_active(ntfn_ptr, badge);
+            mask_ack_bail(irq);
+            UNREACHABLE();
+        }
+
+        idle = true;
+        break;
+    case NtfnState_Waiting:
+        dest = TCB_PTR(notification_ptr_get_ntfnQueue_head(ntfn_ptr));
+        break;
+    default:
+        fail("Invalid notification state");
+    }
+
+    /* at this point we know we're going to have to switch threads */
+    assert(dest != NULL);
+
+     /* Get the bound SC of the signalled thread */
+    sc = dest->tcbSchedContext;
+
+    /* If the signalled thread doesn't have a bound SC, check if one can be
+     * donated from the notification. If not, go to the slowpath */
+    if (!sc || dest->tcbSchedContext->scRefillMax == 0) {
+        sc = SC_PTR(notification_ptr_get_ntfnSchedContext(ntfn_ptr));
+        if (sc == NULL || sc->scTcb != NULL) {
+            slowpath_irq();
+            UNREACHABLE();
+        }
+
+        /* Slowpath the case where dest has its FPU context in the FPU of a core*/
+#if defined(ENABLE_SMP_SUPPORT) && defined(CONFIG_HAVE_FPU)
+        if (nativeThreadUsingFPU(dest)) {
+            slowpath_irq();
+            UNREACHABLE();
+        }
+#endif
+    }
+
+#ifdef CONFIG_ENABLE_SMP
+    if (dest->tcbAffinity != getCurrentCPUIndex()) {
+        slowpath_irq();
+        UNREACHABLE();
+    }
+#endif
+
+    if (sc->scRefillMax > 0) {
+        if (!(refill_ready(sc) && refill_sufficient(sc, 0))) {
+            slowpath_irq();
+        }
+        schedulable = true;
+    }
+
+    /* Get destination thread.*/
+    newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
```

Comment:
Do you need to do any of this stuff if dest isn't the new highest prio thread on the core? Maybe it would be better to only do this stuff conditionally.